indico/modules/events/timetable/util.py

```diff
@@ -0,0 +1,158 @@
+    """Retrieve time blocks that fall within an specific time interval
+       for a given set of categories.
+
+       :param categ_ids: iterable containing list of category IDs
+       :param start_dt: start of search interval (``datetime``, expected
+                        to be in display timezone)
+       :param end_dt: end of search interval (``datetime`` in expected
+                      to be in display timezone)
+       :param detail_level: the level of detail of information
+              (``event|session|contribution``)
+       :param tz: the ``timezone`` information should be displayed in
+       :param from_categ: ``Category`` that will be taken into account to calculate
+                          visibility
+       :returns: a dictionary containing timetable information is a
+                 structured way. See source code for examples.
+    """
+    day_start = start_dt.astimezone(utc)
+    day_end = end_dt.astimezone(utc)
+    dates_overlap = lambda t: (t.start_dt >= day_start) & (t.start_dt <= day_end)
+
+    items = defaultdict(lambda: defaultdict(list))
+
+    # first of all, query TimetableEntries/events that fall within
+    # specified range of dates (and category set)
+    for eid, tt_start_dt in _query_events(categ_ids, day_start, day_end):
+        if tt_start_dt:
+            items[eid][tt_start_dt.astimezone(tz).date()].append(tt_start_dt)
+        else:
+            items[eid] = None
+
+    # then, retrieve detailed information about the events
+    event_ids = set(items)
+    query = (Event.find(Event.id.in_(event_ids))
+             .options(subqueryload(Event.person_links).joinedload(EventPersonLink.person),
+                      joinedload(Event.own_room).noload('owner'),
+                      joinedload(Event.own_venue)))
+
+    scheduled_events = defaultdict(list)
+    ongoing_events = []
+    for e in query:
+        if from_categ and not is_visible_from(e, from_categ):
+            continue
+        local_start_dt = e.start_dt.astimezone(tz).date()
+        local_end_dt = e.end_dt.astimezone(tz).date()
+        if items[e.id] is None:
+            # if there is no TimetableEntry, this means the event has not timetable on that interval
+            for day in iterdays(max(start_dt.date(), local_start_dt), min(end_dt.date(), local_end_dt)):
+                # if the event starts on this date, we've got a time slot
+                if day.date() == local_start_dt:
+                    scheduled_events[day.date()].append((e.start_dt, e))
+                else:
+                    ongoing_events.append(e)
+        else:
+            for start_d, start_dts in items[e.id].viewitems():
+                scheduled_events[start_d].append((start_dts[0], e))
+
+    # result['events'][date(...)] -> [(datetime(....), Event(...))]
+    # result[event_id]['contribs'][date(...)] -> [(TimetableEntry(...), Contribution(...))]
```

Comment:
Event IDs are always ints, so, there should never be collisions. I agree it's not pretty, but I couldn't find a decent key name.